client-lib/nsa.js

```diff
@@ -0,0 +1,210 @@
+const deepAssign = require('deep-assign');
+
+const emitter = require('./emitter');
+const formatTime = require('./format-time');
+const sendToAddon = require('./send-to-addon');
+const sendMetricsEvent = require('./send-metrics-event');
+
+const YtCtrl = require('./ctrls/yt-ctrl');
+const AudioCtrl = require('./ctrls/audio-ctrl');
+const VideoCtrl = require('./ctrls/video-ctrl');
+
+let currentStep, playerMap = {}; // eslint-disable-line no-unused-vars
+
+function step() {
+  let currentTime = 0;
+  let progress = window.AppData.progress;
+  let duration = 0;
+
+  if (playerMap[window.AppData.player]) {
+    currentTime = playerMap[window.AppData.player].time;
+    progress = (currentTime / playerMap[window.AppData.player].duration) || window.AppData.progress;
+    duration = playerMap[window.AppData.player].duration;
+  }
+
+  window.AppData = deepAssign(window.AppData, {
+    currentTime: currentTime,
+    time: `${formatTime(currentTime)} / ${formatTime(duration)}`,
+    progress: progress,
+    duration: duration
+  });
+
+  if (currentTime >= window.AppData.duration) {
+    window.AppData = deepAssign(window.AppData, {
+      playing: false,
+      exited: currentTime ? true : false
+    });
+
+    sendMetricsEvent('player_view', 'video_ended');
+  }
+
+  if (window.AppData.playing) currentStep = requestAnimationFrame(step);
+}
+
+emitter.on('reset', () => {
+  currentStep = null;
+  Object.keys(playerMap).forEach((k) => {
+    if (playerMap[k]) playerMap[k].pause();
+  });
+
+  deepAssign(window.AppData, {
+    currentTime: 0,
+    playing: false,
+    error: false,
+    exited: false,
+    time: `${formatTime(0)} / ${formatTime(window.AppData.duration)}`,
+    progress: 0.001
+  });
+})
+
+emitter.on('init', (opts) => {
+  sendMetricsEvent('player_view', 'init');
+
+  if (window.AppData.player === 'audio') {
+    playerMap['audio'] = new AudioCtrl(opts);
```

Comment:
Related suggestion: seems like you could remove the last couple of `window.AppData` get/set calls from the components and just use emitters there instead. Looks like just a couple of instances in the player and progress components. I think it'd make more sense to have a hard rule than to sometimes use emitters and sometimes update state directly. Plus, again, being consistent about emitter usage simplifies unit testing components: you'd just mock out the emitter library.